### DISCUSSION TITLE
extend ModelQueryBuilderContract?

### DISCUSSION THREAD

**Discussion by varun-nambiar** [adonisjs/core#2766] [Ideas]
Is it possible to do something like this with ModelQueryBuilderContract?

https://docs.adonisjs.com/guides/response#informing-typescript-about-the-method

I tried creating the following
![image](https://user-images.githubusercontent.com/1249268/121420250-77f2c200-c98a-11eb-89cb-b8e46d9fc60d.png)

I'm trying to add a new method "archive" (very similar to delete)

I am able to add methods and hooks into models and achieve this for individual records (or map for each record)

```javascript
const usersA = User.query();
usersA.where('status', 'inactive');
// delete from users where status ='inactive';
await usersA.delete();

const usersB = User.query();
usersB.where('status', 'inactive');
// update users set deleted = 1 where status ='inactive';
await usersB.archive();
```

Is this possible by creating a contract and then creating a DatabaseProvider (similar to what is in the guide for Response/Request contract)?

❤️

**Comment by varun-nambiar** [comment]
**Update**

I got this to work now. Heres my DatabaseProvider... but
```javascript
import { ApplicationContract } from '@ioc:Adonis/Core/Application'

export default class DatabaseProvider {
  constructor(protected app: ApplicationContract) {
  }

  public register() {
    // Register your own bindings
    this.app.container.bind('Binding', () => { })
  }

  public async boot() {
    const ModelQueryBuilder = (await import('@ioc:Adonis/Lucid/Database')).default.ModelQueryBuilder
    ModelQueryBuilder.macro('archive', function () {
      // has deleted column defined?
      if (this.model.$hasColumn('deleted')) {
        return this.update({
          deleted: 1
        })
      } else {
        return this.delete()
      }
    })
  }
}

```

Its working as expected (so far)  
I still get an error on my database contract  

_contract/database.ts_
```javascript
declare module '@ioc:Adonis/Lucid/Model' {
  interface ModelQueryBuilderContract {
    archive(): this
  }
}
```

**Error message:**  
All declarations of 'ModelQueryBuilderContract' must have identical type parameters.


I feel like I'm very much close to get it working or I may be way too off.
Thoughts/suggestions/guide any one please.

@RomainLanz @thetutlage 

Thank you!

**Comment by varun-nambiar** [comment]
**Update**
I updated the packages and got this working!

Had to change '@ioc:Adonis/Lucid/Model' to '@ioc:Adonis/Lucid/Orm'

**Reply by thetutlage** [reply]
There is a release on Monday that also includes the docs for extending the query builder

**Reply by epenflow** [reply]
any example extending ModelQueryBuilder in adonis js v6?
